fingerprint.c

```diff
@@ -322,24 +359,10 @@ static int fingerprint_cancel(struct fingerprint_device *dev)
     sony_fingerprint_device_t *sdev = (sony_fingerprint_device_t*)dev;
     fingerprint_notify_t callback = sdev->device.notify;
 
-    pthread_mutex_lock(&sdev->lock);
-    bool thread_running = sdev->worker.thread_running ;
-    pthread_mutex_unlock(&sdev->lock);
-
-    ALOGI("%s : check thread running",__func__);
-    if (!sdev->worker.thread_running ) {
-        ALOGI("%s : - (thread not running)",__func__);
-        return 0;
+    if (!setState(sdev, STATE_IDLE)){
```

Comment:
random thought - maybe it's better to have an explicit cancel bool instead of using state?
That way you can signal cancel, thread checks it, and you can loop until state == IDLE before returning.
This should fix any potential enroll -> cancel -> auth race conditions

Reply:
@spiiroin agreed it should wait after calling STATE_IDLE before returning